web/modules/custom/lit_cover_service/src/Service/BatchService.php

```diff
@@ -0,0 +1,252 @@
+      $context['results'][] = $node->nid;
+
+      // Update our progress information.
+      $context['sandbox']['progress']++;
+      $context['sandbox']['current_node'] = $node->nid;
+    }
+
+    self::deleteFileEntities($deletedImageFiles);
+
+    // Inform the batch engine that we are not finished,
+    // and provide an estimation of the completion level we reached.
+    if ($context['sandbox']['progress'] != $context['sandbox']['max']) {
+      $context['finished'] = $context['sandbox']['progress'] / $context['sandbox']['max'];
+    }
+  }
+
+  /**
+   * Batch fetch process callback.
+   *
+   * @param int $batchId
+   *   Id of the batch
+   * @param $batchTotal
+   *   Total number of batch operations
+   * @param array $nids
+   *   The nids to process
+   * @param object $context
+   *   Context for operations
+   *
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   */
+  public function fetchBookCovers($batchId, $batchTotal, $nids, &$context) {
+    if (!isset($context['sandbox']['progress'])) {
+      $context['sandbox']['progress'] = 0;
+      $context['sandbox']['current_node'] = 0;
+      $context['sandbox']['max'] = $batchTotal;
+    }
+
+    $context['message'] = t('Fetching batch @batchId of @batchTotal', [
+      '@batchId' => $batchId,
+      '@batchTotal' => $batchTotal,
+    ]);
+
+    $nodes = Node::loadMultiple($nids);
+    foreach ($nodes as $node) {
+      self::setImageFile($node);
+      $node->save();
+
+      // Store some result for post-processing in the finished callback.
+      $context['results'][] = $node->nid;
+
+      // Update our progress information.
+      $context['sandbox']['progress']++;
+      $context['sandbox']['current_node'] = $node->nid;
+    }
+
+    // Inform the batch engine that we are not finished,
+    // and provide an estimation of the completion level we reached.
+    if ($context['sandbox']['progress'] != $context['sandbox']['max']) {
+      $context['finished'] = $context['sandbox']['progress'] / $context['sandbox']['max'];
+    }
+  }
+
+  /**
+   * Replace process callback.
+   *
+   * @param int $batchId
+   *   Id of the batch
+   * @param $batchTotal
+   *   Total number of batch operations
+   * @param array $nids
+   *   The nids to process
+   * @param object $context
+   *   Context for operations
+   *
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   */
+  public function replaceBookCovers($batchId, $batchTotal, $nids, &$context) {
+    if (!isset($context['sandbox']['progress'])) {
+      $context['sandbox']['progress'] = 0;
+      $context['sandbox']['current_node'] = 0;
+      $context['sandbox']['max'] = $batchTotal;
+    }
+
+    $context['message'] = t('Fetching batch @batchId of @batchTotal', [
+      '@batchId' => $batchId,
+      '@batchTotal' => $batchTotal,
+    ]);
+
+    $nodes = Node::loadMultiple($nids);
+    $deletedImageFiles = [];
+    foreach ($nodes as $node) {
+      $deletedImageFiles[] = self::clearImageField($node);
```

Comment:
`clearImageField` set the field to `null` will this actually delete the local file? I known that you call `deleteFileEntities` later on, but what if the first 50 nodes are cleared and the 51 fails with an exception?

Reply:
See above